### ISSUE TITLE
OptionAttribute IEnumerable<T> possible unintended behaviour

### ISSUE THREAD

**Issue by ericnewton76** [commandlineparser/commandline#57]
<a href="https://github.com/usagirei"><img src="https://avatars3.githubusercontent.com/u/12700106?v=4" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [usagirei](https://github.com/usagirei)**
_Wednesday Jan 06, 2016 at 21:36 GMT_
_Originally opened as https://github.com/gsscoder/commandline/issues/277_

----

Given the class

``` csharp
public class Options{
    private IEnumerable<string> _inputDirs;

    [Option('c', "compress",
        HelpText = "Compress Match Pattern, Pipe Separated (|) ",
        Separator = '|',
        Max = 1,
        Default = new[]
        {
            "*.txt", "*.log", "*.ini"
        })]
    public IEnumerable<string> Compress { get; set; }

    [Value(0,
        HelpText = "Input Directories.",
        Min = 1,
        Required = true)]
    public IEnumerable<string> InputDirs
    {
        get { return _inputDirs; }
        set { _inputDirs = value.SelectMany(Globbing.GlobDirs).ToList(); }
    }

    [Option('n', "name",
        HelpText = "Metadata Name.",
        Default = WILDCARD)]
    public string Name { get; set; }
}
```

The following happens:
- If the passed argument is: 
  `"*" -c "*.*"`
  It'll work as intended
- If the passed argument is: 
  `"*" -c "*"` or `a b c -c a` or `a b c -c a -n b`
  Arguments present in the _options_ are removed from the _value_ arguments. However two options can have the same value with no problems.
  In the first case no argument will be passed to the Value.
  In the second, only b and c.
  In the third only c
- If the passed argument is:
  `-c "*" "*.*"`
  Compress will receive the two arguments, even if Min and Max are set, it'll consume anything until the next **option** is found (values get absorbed). IEnumerable<T> is ignoring the Max property, same without Separator property.
- If the passed argument is:
  `-c "*.log|*.txt" "*"`
  Compress will receive three arguments, the first one split, plus anything until the next _option_, meaning you can't have a value after a option IEnumerable is set. Same as above.
- If the passed argument is:
  `-c "*" "*.log|*.txt"`
  Compress will receive _two_ arguments, the second one will _not_ get split, meaning the second argument being read is probably a unintended behaviour, shouldn't a Sequence IEnumerable read only one argument?

Tested with _2.0.275-beta_

TL;DR: IEnumerable<T> Ignores Max Property, "duplicate" attributes present in a _Option_ Property gets removed from a _Value_ Property

**Comment by ericnewton76** [comment]
<a href="https://github.com/Mizipzor"><img src="https://avatars1.githubusercontent.com/u/139879?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [Mizipzor](https://github.com/Mizipzor)**
_Friday Jan 29, 2016 at 21:58 GMT_

----

Great findings! Would you like to give it a go writing unit tests that cover these cases?

**Comment by ericnewton76** [comment]
<a href="https://github.com/Mizipzor"><img src="https://avatars1.githubusercontent.com/u/139879?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [Mizipzor](https://github.com/Mizipzor)**
_Tuesday Feb 02, 2016 at 20:47 GMT_

----

Decided I wanted to try this one out, but I'm not sure what `Globbing` is. Is that another assembly you are referencing?

**Comment by ericnewton76** [comment]
<a href="https://github.com/usagirei"><img src="https://avatars3.githubusercontent.com/u/12700106?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [usagirei](https://github.com/usagirei)**
_Wednesday Feb 03, 2016 at 22:11 GMT_

----

Yes, it's a method that will expand a string into an IEnumerable<string> of all globbed files (file_*.txt) 

---

Edit: I can try my hand at a few tests that will likely fail, sorry for the delayed response

---

Edit2: I'm failing to compile the solution, it's missing the nuget package entries (recoverable, not a big deal) but the tests won't compile since it's missing a reference to CSharpx, which is not available on the solution.

Here's the Globbing class in question, if you require it https://gist.github.com/usagirei/7fee07ed17416f7f5690

I may be missing the stuff in order to build it

**Comment by ericnewton76** [comment]
<a href="https://github.com/Mizipzor"><img src="https://avatars1.githubusercontent.com/u/139879?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [Mizipzor](https://github.com/Mizipzor)**
_Thursday Feb 04, 2016 at 08:09 GMT_

----

All nuget dependencies should be automatically downloaded if they are missing. How are you building?

**Comment by ericnewton76** [comment]
<a href="https://github.com/usagirei"><img src="https://avatars3.githubusercontent.com/u/12700106?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [usagirei](https://github.com/usagirei)**
_Thursday Feb 04, 2016 at 12:19 GMT_

----

Normally, but they didn't restore properly, had to manually re-install FluentAssertions and xUnit Packages, however CSharpx is still missing

EDIT: I think i know what didn't work had to run the external build.cmd file before trying to build trough visual studio, still I'm not familiar with this testing suite

EDIT2: Got it working, will try to write some tests for it, got some of the situations failing already, will try to get them up and running when i got some more spare time

EDIT3: These should cover these issues i think usagirei/commandline@c4855e8
(Tests are on usagirei/commandline@93014b6a but i forgot to change the filenames, then to save the project...)
